tower-util/src/combinators/with.rs

```diff
@@ -0,0 +1,70 @@
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pub struct With<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> With<S, F> {
+    pub fn new(inner: S, f: F) -> Self {
+        With { inner, f }
+    }
+}
+
+impl<S, F, NewRequest, OldRequest> Service<NewRequest> for With<S, F>
+where
+    S: Service<OldRequest>,
+    F: FnOnce(NewRequest) -> OldRequest,
+    F: Clone,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), S::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: NewRequest) -> S::Future {
+        self.inner.call((self.f.clone())(request))
+    }
+}
+
+#[derive(Debug)]
+pub struct WithLayer<F, OldRequest, NewRequest> {
+    f: F,
+    old_request: PhantomData<OldRequest>,
+    new_request: PhantomData<NewRequest>,
```

Comment:
nit/TIOLI: I think we wouldn't need any `PhantomData` here if we just changed this to 
```suggestion
pub struct WithLayer<F, OldRequest, NewRequest>
where
   F: FnOnce(NewRequest) -> OldRequest,
{
    f: F,
```
but then we would need to put this bound on all impls for this type...

Reply:
I remember noticing this at the time but opted for the consistency of having the same layout of the phantom data on all these combinators and fewer impls.